logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -252,15 +252,21 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
         executeAgeRetentionPolicy();
 
         if (currentQueueSize.longValue() + eventPayloadSize > maxQueueSize) {
-            if (storageType == QueueStorageType.DROP_NEWER) {
-                lastError = String.format("Cannot write event to DLQ(path: %s): reached maxQueueSize of %d", queuePath, maxQueueSize);
-                logger.error(lastError);
-                droppedEvents.add(1L);
-                return;
-            } else {
-                do {
-                    dropTailSegment();
-                } while (currentQueueSize.longValue() + eventPayloadSize > maxQueueSize);
+            // load currentQueueSize from filesystem because there could be a consumer
+            // that's already cleaning
+            this.currentQueueSize.set(computeQueueSize());
+
+            if (currentQueueSize.longValue() + eventPayloadSize > maxQueueSize) {
+                if (storageType == QueueStorageType.DROP_NEWER) {
+                    lastError = String.format("Cannot write event to DLQ(path: %s): reached maxQueueSize of %d", queuePath, maxQueueSize);
+                    logger.error(lastError);
+                    droppedEvents.add(1L);
+                    return;
+                } else {
+                    do {
+                        dropTailSegment();
+                    } while (currentQueueSize.longValue() + eventPayloadSize > maxQueueSize);
```

Comment:
Should we use `currentQueueSize` here? Or at all? We could end up dropping more tail segments than is necessary